soc/arm/ti_simplelink/cc13x2_cc26x2/power.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2019 Linaro Limited.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <init.h>
+#include <power/power.h>
+
+#include <ti/drivers/Power.h>
+#include <ti/drivers/power/PowerCC26X2.h>
+
+#include <ti/drivers/dpl/ClockP.h>
+
+#include <ti/devices/cc13x2_cc26x2/driverlib/cpu.h>
+#include <ti/devices/cc13x2_cc26x2/driverlib/vims.h>
+#include <ti/devices/cc13x2_cc26x2/driverlib/sys_ctrl.h>
+
+#include <logging/log.h>
+#define LOG_LEVEL CONFIG_SOC_LOG_LEVEL
+LOG_MODULE_REGISTER(soc);
+
+/* Configuring TI Power module to not use its policy function (we use Zephyr's
+ * instead), and disable oscillator calibration functionality for now.
+ */
+const PowerCC26X2_Config PowerCC26X2_config = {
+	.policyInitFxn      = NULL,
+	.policyFxn          = NULL,
+	.calibrateFxn       = NULL,
+	.enablePolicy       = false,
+	.calibrateRCOSC_LF  = false,
+	.calibrateRCOSC_HF  = false
+};
+
+extern PowerCC26X2_ModuleState PowerCC26X2_module;
+
+/*
+ * Power state mapping:
+ * SYS_POWER_STATE_SLEEP_1: Idle
+ * SYS_POWER_STATE_SLEEP_2: Standby
+ */
+
+/* Invoke Low Power/System Off specific Tasks */
+void sys_set_power_state(enum power_states state)
+{
+#ifdef CONFIG_SYS_POWER_SLEEP_STATES
+#ifdef CONFIG_HAS_SYS_POWER_STATE_SLEEP_1
+	u32_t modeVIMS;
+	u32_t constraints;
+#endif
+#endif
+
+	LOG_DBG("SoC entering power state %d", state);
+
+	/* Switch to using PRIMASK instead of BASEPRI register, since
+	 * we are only able to wake up from standby while using PRIMASK.
+	 */
+	/* Set PRIMASK */
+	CPUcpsid();
+	/* Set BASEPRI to 0 */
+	irq_unlock(0);
+
+	switch (state) {
+#ifdef CONFIG_SYS_POWER_SLEEP_STATES
+#ifdef CONFIG_HAS_SYS_POWER_STATE_SLEEP_1
+	case SYS_POWER_STATE_SLEEP_1:
+		/* query the declared constraints */
+		constraints = Power_getConstraintMask();
+		/* 1. Get the current VIMS mode */
+		do {
+			modeVIMS = VIMSModeGet(VIMS_BASE);
+		} while (modeVIMS == VIMS_MODE_CHANGING);
+
+		/* 2. Configure flash to remain on in IDLE or not and keep
+		 *    VIMS powered on if it is configured as GPRAM
+		 * 3. Always keep cache retention ON in IDLE
+		 * 4. Turn off the CPU power domain
+		 * 5. Ensure any possible outstanding AON writes complete
+		 * 6. Enter IDLE
+		 */
+		if ((constraints & (1 << PowerCC26XX_NEED_FLASH_IN_IDLE)) ||
+			(modeVIMS == VIMS_MODE_DISABLED)) {
+			SysCtrlIdle(VIMS_ON_BUS_ON_MODE);
+		} else {
+			SysCtrlIdle(VIMS_ON_CPU_ON_MODE);
+		}
+
+		/* 7. Make sure MCU and AON are in sync after wakeup */
+		SysCtrlAonUpdate();
+		break;
+#endif /* CONFIG_HAS_SYS_POWER_STATE_SLEEP_1 */
+
+#ifdef CONFIG_HAS_SYS_POWER_STATE_SLEEP_2
+	case SYS_POWER_STATE_SLEEP_2:
+		/* schedule the wakeup event */
+		ClockP_start(ClockP_handle((ClockP_Struct *)
+			&PowerCC26X2_module.clockObj));
+
+		/* go to standby mode */
+		Power_sleep(PowerCC26XX_STANDBY);
```

Comment:
sorry I didn't understand below code clearly, could you please give some explain about this state
```
            if (earlyPG) {
                PowerCC32XX_enterLPDS(PRCMLPDSEnter);
            }
            else {
                PowerCC32XX_enterLPDS(ROM_PRCMLPDSEnterDirect);
            }
```

Reply:
I think you may be looking at the wrong implementation of Power_sleep(). The code you are referring to seems to be targeted for CC32xx, which is a different family of MCUs. Power_sleep() for cc13x2/cc26x2 is in zephyrproject/modules/hal/ti/simplelink/source/ti/drivers/power/PowerCC26X2.c.

Reply:
@vanti thanks. Review the right code, have some questions: 1) in SysCtrlStandby, if retainCache is false, how CPU restore its content; 2) it seems all power domains are turned off when invoke PRCMDeepSleep, what the resume process is doing in the hardware? I think it's better to add more comments for the resume.

Reply:
@wentongwu 
1. If the VIMS was in cache mode before entering standby, upon resume the code in Power_sleep() would bring the VIMS back to cache mode:

```
            if (retainCache == false) {

                /* 5.1 If previously in a cache mode, restore the mode now */
                if (modeVIMS == VIMS_MODE_ENABLED) {
                    VIMSModeSet(VIMS_BASE, modeVIMS);
                }

```

The cache is invalidated/initialized automatically when entering cache mode, per section 8.3.2.1 of the TRM (http://www.ti.com/lit/ug/swcu185d/swcu185d.pdf). The content does not need to be restored in the traditional sense, since this is acting as a cache - it'll be repopulated over time.

2. The resume process is documented in Power_sleep() from source/ti/drivers/power/PowerCC26X2.c in the HAL. PRCMDeepSleep() only causes the CPU to enter standby mode. If you look at the code of SysCtrlStandby(), which is invoked by Power_sleep(), you would see that power domains are turned off prior to calling PRCMDeepSleep():
 ```
        // Request power off of domains in the MCU voltage domain
        PRCMPowerDomainOff(PRCM_DOMAIN_RFCORE | PRCM_DOMAIN_SERIAL | PRCM_DOMAIN_PERIPH | PRCM_DOMAIN_CPU); 

...
        // Invoke deep sleep to go to STANDBY
        PRCMDeepSleep()
```

During resume, the power domains are turned on with an explicit function call in Power_sleep():

```
            /* 6. Start re-powering power domains */
            PRCMPowerDomainOn(poweredDomains);
```

Reply:
@wentongwu 
1. If the VIMS was in cache mode before entering standby, upon resume the code in Power_sleep() would bring the VIMS back to cache mode:

```
            if (retainCache == false) {

                /* 5.1 If previously in a cache mode, restore the mode now */
                if (modeVIMS == VIMS_MODE_ENABLED) {
                    VIMSModeSet(VIMS_BASE, modeVIMS);
                }

```

The cache is invalidated/initialized automatically when entering cache mode, per section 8.3.2.1 of the TRM (http://www.ti.com/lit/ug/swcu185d/swcu185d.pdf). The content does not need to be restored in the traditional sense, since this is acting as a cache - it'll be repopulated over time.

2. The resume process is documented in Power_sleep() from _source/ti/drivers/power/PowerCC26X2.c_ in the HAL. PRCMDeepSleep() only causes the CPU to enter standby mode. If you look at the code of SysCtrlStandby(), which is invoked by Power_sleep(), you would see that power domains are turned off prior to calling PRCMDeepSleep() and not while invoking it:
 ```
        // Request power off of domains in the MCU voltage domain
        PRCMPowerDomainOff(PRCM_DOMAIN_RFCORE | PRCM_DOMAIN_SERIAL | PRCM_DOMAIN_PERIPH | PRCM_DOMAIN_CPU); 

...
        // Invoke deep sleep to go to STANDBY
        PRCMDeepSleep()
```

During resume, the power domains are then turned on with an explicit function call in Power_sleep():

```
            /* 6. Start re-powering power domains */
            PRCMPowerDomainOn(poweredDomains);
```